# Case Study: Automated Machine Learning (AutoML) for Autonomous Intrusion Detection System Development
This is the code for the paper entitled "**[Enabling AutoML for Zero-Touch Network Security: Use-Case Driven Analysis](https://ieeexplore.ieee.org/document/10472316)**" published in *IEEE Transactions on Network and Service Management* (IF:5.3).<br>
Authors: Li Yang (liyanghart@gmail.com), Mirna El Rajab, Abdallah Shami, and Sami Muhaidat<br>

L. Yang, M. E. Rajab, A. Shami, and S. Muhaidat, "Enabling AutoML for Zero-Touch Network Security: Use-Case Driven Analysis," IEEE Transactions on Network and Service Management, pp. 1-28, 2024, doi: https://doi.org/10.1109/TNSM.2024.3376631.

# Code Part 3: Adversarial Machine Learning (AML) Attack and Defense
As many network services and functionalities rely on AI/ML models, they are becoming increasingly vulnerable to AML attacks. AML attacks exploit the weaknesses and vulnerabilities of ML models by generating adversarial inputs that can deceive or manipulate the models into making incorrect predictions. In networks, AML attacks pose a significant threat to overall network security and reliability.  
This case study aims to demonstrate the detrimental impact that AML attacks can have on ML models in networks and presents basic defense strategies to mitigate these attacks, thereby ensuring the accuracy of the ML-based IDS. In this case study, three common types of adversarial attacks ((i.e., DTA, FGSM, and BIM)) are used to generate adversarial samples to probe the vulnerability of the IDS.  
Subsequently, basic defense mechanisms, including adversarial sample detection and filtering, are devised to safeguard the AutoML-based IDS against AML attacks.  

## Dataset 1: CICIDS2017
A subset of the network traffic data randomly sampled from the [CICIDS2017 dataset](https://www.unb.ca/cic/datasets/ids-2017.html).  

The Canadian Institute for Cybersecurity Intrusion Detection System 2017 (CICIDS2017) dataset has the most updated network threats. The CICIDS2017 dataset is close to real-world network data since it has a large amount of network traffic data, a variety of network features, various types of attacks, and highly imbalanced classes.

## Import libraries

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split,cross_val_score
import lightgbm as lgb
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from scipy.stats import shapiro
from imblearn.over_sampling import SMOTE
import time

In [3]:
import warnings
warnings.filterwarnings('ignore')

## Read the sampled CICIDS2017 dataset

In [4]:
df = pd.read_csv("/content/cic_0.01km.csv")

In [5]:
df

Flow Duration  Total Length of Fwd Packets  Fwd Packet Length Max  \
0              50833                            0                      0   
1                 49                            0                      0   
2                306                            6                      6   
3              63041                           65                     65   
4              47682                           43                     43   
...              ...                          ...                    ...   
28298             45                            0                      0   
28299      114309573                          511                    427   
28300          48850                           80                     40   
28301            260                           66                     33   
28302             25                            6                      6   

       Fwd Packet Length Mean  Bwd Packet Length Max  Bwd Packet Length Min  \
0                      0.0000                      0                      0   
1                      0.0000                      0                      0   
2                      6.0000                      6                      6   
3                     65.0000                    124                    124   
4                     43.0000                     59                     59   
...                       ...                    ...                    ...   
28298                  0.0000                      0                      0   
28299                 31.9375                    746                      0   
28300                 40.0000                     72                     72   
28301                 33.0000                     97                     97   
28302                  6.0000                      6                      6   

       Flow IAT Mean  Flow IAT Min  Fwd IAT Min  Fwd Header Length  \
0       5.083300e+04         50833            0                 32   
1       4.900000e+01            49           49                 64   
2       3.060000e+02           306            0                 20   
3       6.304100e+04         63041            0                 32   
4       4.768200e+04         47682            0                 32   
...              ...           ...          ...                ...   
28298   4.500000e+01            45            0                 32   
28299   3.941709e+06            94          165                332   
28300   1.628333e+04             1           48                 64   
28301   8.666667e+01            48           48                 40   
28302   2.500000e+01            25            0                 20   

       Bwd Header Length  Fwd Packets/s  Bwd Packets/s  Min Packet Length  \
0                     32      19.672260      19.672260                  0   
1                      0   40816.326530       0.000000                  0   
2                     20    3267.973856    3267.973856                  6   
3                     32      15.862693      15.862693                 65   
4                     32      20.972275      20.972275                 43   
...                  ...            ...            ...                ...   
28298                 32   22222.222220   22222.222220                  0   
28299                424       0.139971       0.122474                  0   
28300                 64      40.941658      40.941658                 40   
28301                 40    7692.307692    7692.307692                 33   
28302                 20   40000.000000   40000.000000                  6   

       URG Flag Count  Down/Up Ratio  Init_Win_bytes_forward  \
0                   1              1                     319   
1                   0              0                     277   
2                   0              1                       0   
3                   0              1                      -1   
4                   0              1                      -

# 1. Automated Data Pre-Processing

## Automated Transformation/Encoding
Automatically identify and transform string/text features into numerical features to make the data more readable by ML models

In [6]:
# Define the automated data encoding function
def Auto_Encoding(df):
    cat_features=[x for x in df.columns if df[x].dtype=="object"] ## Find string/text features
    le=LabelEncoder()
    for col in cat_features:
        if col in df.columns:
            i = df.columns.get_loc(col)
            # Transform to numerical features
            df.iloc[:,i] = df.apply(lambda i:le.fit_transform(i.astype(str)), axis=0, result_type='expand')
    return df

In [7]:
df=Auto_Encoding(df)

## Automated Imputation
Detect and impute missing values to improve data quality

In [8]:
# Define the automated data imputation function
def Auto_Imputation(df):
    if df.isnull().values.any() or np.isinf(df).values.any(): # if there is any empty or infinite values
        df.replace([np.inf, -np.inf], np.nan, inplace=True)
        df.fillna(0, inplace = True)  # Replace empty values with zeros; there are other imputation methods discussed in the paper
    return df

In [9]:
df=Auto_Imputation(df)

## Automated normalization
Normalize the range of features to a similar scale to improve data quality

In [10]:
def Auto_Normalization(df):
    stat, p = shapiro(df)
    print('Statistics=%.3f, p=%.3f' % (stat, p))
    # interpret
    alpha = 0.05
    numeric_features = df.drop(['Label'],axis = 1).dtypes[df.dtypes != 'object'].index

    # check if the data distribution follows a Gaussian/normal distribution
    # If so, select the Z-score normalization method; otherwise, select the min-max normalization
    # Details are in the paper
    if p > alpha:
        print('Sample looks Gaussian (fail to reject H0)')
        df[numeric_features] = df[numeric_features].apply(
            lambda x: (x - x.mean()) / (x.std()))
        print('Z-score normalization is automatically chosen and used')
    else:
        print('Sample does not look Gaussian (reject H0)')
        df[numeric_features] = df[numeric_features].apply(
            lambda x: (x - x.min()) / (x.max()-x.min()))
        print('Min-max normalization is automatically chosen and used')
    return df

In [11]:
df=Auto_Normalization(df)

Statistics=0.076, p=0.000
Sample does not look Gaussian (reject H0)
Min-max normalization is automatically chosen and used


In [12]:
df=Auto_Imputation(df)

## Train-test split
Split the dataset into the training and the test set

In [13]:
X = df.drop(['Label'],axis=1)
y = df['Label']

# Here we used the 80%/20% split, it can be changed based on specific tasks
#X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2, shuffle=False,random_state = 0)
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2,random_state = 0)

## Automated data balancing
Generate minority class samples to solve class-imbalance and improve data quality.  
Synthetic Minority Over-sampling Technique (SMOTE) method is used.

In [14]:
pd.Series(y_train).value_counts()

Label
0    18126
1     4516
Name: count, dtype: int64

In [20]:
# For binary data (can be modified for multi-class data with the same logic)
def Auto_Balancing(X_train, y_train):
    number0 = pd.Series(y_train).value_counts().iloc[0]
    number1 = pd.Series(y_train).value_counts().iloc[1]

    if number0 > number1:
        nlarge = number0
    else:
        nlarge = number1

    # evaluate whether the incoming dataset is imbalanced (the abnormal/normal ratio is smaller than a threshold (e.g., 50%))
    if (number1/number0 > 1.5) or (number0/number1 > 1.5):
        smote=SMOTE(sampling_strategy={0:nlarge, 1:nlarge})
        X_train, y_train = smote.fit_resample(X_train, y_train)

    return X_train, y_train

In [21]:
X_train, y_train = Auto_Balancing(X_train, y_train)

In [22]:
pd.Series(y_train).value_counts()

Label
0    18126
1    18126
Name: count, dtype: int64

# 2. Automated Feature Engineering
Feature selection method 1: **Information Gain (IG)**, used to remove irrelevant features to improve model efficiency  
Feature selection method 2: **Pearson Correlation**, used to remove redundant features to improve model efficiency and accuracy  

In [23]:
# Remove irrelevant features and select important features
def Feature_Importance_IG(data):
    features = data.drop(['Label'],axis=1).values  # "Label" should be changed to the target class variable name if different
    labels = data['Label'].values

    # Extract feature names
    feature_names = list(data.drop(['Label'],axis=1).columns)

    # Empty array for feature importances
    feature_importance_values = np.zeros(len(feature_names))
    model = lgb.LGBMRegressor(verbose = -1)
    model.fit(features, labels)
    feature_importances = pd.DataFrame({'feature': feature_names, 'importance': model.feature_importances_})

    # Sort features according to importance
    feature_importances = feature_importances.sort_values('importance', ascending = False).reset_index(drop = True)

    # Normalize the feature importances to add up to one
    feature_importances['normalized_importance'] = feature_importances['importance'] / feature_importances['importance'].sum()
    feature_importances['cumulative_importance'] = np.cumsum(feature_importances['normalized_importance'])

    cumulative_importance=0.90 # Only keep the important features with cumulative importance scores>=90%. It can be changed.

    # Make sure most important features are on top
    feature_importances = feature_importances.sort_values('cumulative_importance')

    # Identify the features not needed to reach the cumulative_importance
    record_low_importance = feature_importances[feature_importances['cumulative_importance'] > cumulative_importance]

    to_drop = list(record_low_importance['feature'])
#     print(feature_importances.drop(['importance'],axis=1))
    return to_drop

In [27]:
# Remove redundant features
def Feature_Redundancy_Pearson(data):
    correlation_threshold=0.90 # Only remove features with the redundancy>90%. It can be changed
    features = data.drop(['Label'],axis=1)
    corr_matrix = features.corr()

    # Extract the upper triangle of the correlation matrix
    upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k = 1).astype(np.bool))

    # Select the features with correlations above the threshold
    # Need to use the absolute value
    to_drop = [column for column in upper.columns if any(upper[column].abs() > correlation_threshold)]

    # Dataframe to hold correlated pairs
    record_collinear = pd.DataFrame(columns = ['drop_feature', 'corr_feature', 'corr_value'])

    # Iterate through the columns to drop
    for column in to_drop:

        # Find the correlated features
        corr_features = list(upper.index[upper[column].abs() > correlation_threshold])

        # Find the correlated values
        corr_values = list(upper[column][upper[column].abs() > correlation_threshold])
        drop_features = [column for _ in range(len(corr_features))]

        # Record the information (need a temp df for now)
        temp_df = pd.DataFrame.from_dict({'drop_feature': drop_features,
                                         'corr_feature': corr_features,
                                         'corr_value': corr_values})
        record_collinear = pd.concat([record_collinear, temp_df], ignore_index=True)
#     print(record_collinear)
    return to_drop

In [25]:
def Auto_Feature_Engineering(df):
    drop1 = Feature_Importance_IG(df)
    dfh1 = df.drop(columns = drop1)

    drop2 = Feature_Redundancy_Pearson(dfh1)
    dfh2 = dfh1.drop(columns = drop2)

    return dfh2

In [28]:
dfh2 = Auto_Feature_Engineering(df)
dfh2

Flow Duration  Total Length of Fwd Packets  Fwd Packet Length Max  \
0       4.236419e-04                     0.000000               0.000000   
1       4.416669e-07                     0.000000               0.000000   
2       2.583334e-06                     0.000008               0.000242   
3       5.253752e-04                     0.000090               0.002619   
4       3.973835e-04                     0.000060               0.001732   
...              ...                          ...                    ...   
28298   4.083335e-07                     0.000000               0.000000   
28299   9.525802e-01                     0.000710               0.017204   
28300   4.071168e-04                     0.000111               0.001612   
28301   2.200001e-06                     0.000092               0.001330   
28302   2.416668e-07                     0.000008               0.000242   

       Fwd Packet Length Mean  Bwd Packet Length Max  Flow IAT Mean  \
0                    0.000000               0.000000   4.707129e-04   
1                    0.000000               0.000000   4.907407e-07   
2                    0.001556               0.000516   2.870370e-06   
3                    0.016856               0.010660   5.837500e-04   
4                    0.011151               0.005072   4.415370e-04   
...                       ...                    ...            ...   
28298                0.000000               0.000000   4.537037e-07   
28299                0.008282               0.064133   3.649735e-02   
28300                0.010373               0.006190   1.508086e-04   
28301                0.008558               0.008339   8.395061e-07   
28302                0.001556               0.000516   2.685185e-07   

       Flow IAT Min   Fwd IAT Min  Fwd Header Length  Bwd Packets/s  \
0      4.707129e-04  0.000000e+00           0.000150   1.311484e-05   
1      4.907407e-07  4.083333e-07           0.000299   0.000000e+00   
2      2.870370e-06  0.000000e+00           0.000094   2.178649e-03   
3      5.837500e-04  0.000000e+00           0.000150   1.057513e-05   
4      4.415370e-04  0.000000e+00           0.000150   1.398152e-05   
...             ...           ...                ...            ...   
28298  4.537037e-07  0.000000e+00           0.000150   1.481481e-02   
28299  9.074074e-07  1.375000e-06           0.001554   8.164962e-08   
28300  4.629629e-08  4.000000e-07           0.000299   2.729444e-05   
28301  4.814815e-07  4.000000e-07           0.000187   5.128205e-03   
28302  2.685185e-07  0.000000e+00           0.000094   2.666667e-02   

       Init_Win_bytes_forward  Init_Win_bytes_backward  Label  
0                    0.004883                 0.002350      0  
1                    0.004242                 0.000000      0  
2                    0.000015                 0.000015      0  
3                    0.000000                 0.000000      0  
4                    0.000000                 0.000000      0  
...                       ...                      ...    ...  
28298                0.005341                 0.004700      0  
28299                0.125015                 0.005249      0  
28300                0.000000                 0.000000      0  
28301                0.000000                 0.000000      0  
28302                0.171768                 0.534073      0  

[28303 rows x 13 columns]

## Data Split & Balancing (After Feature Engineering)

In [29]:
X = dfh2.drop(['Label'],axis=1)
y = dfh2['Label']

#X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2, shuffle=False,random_state = 0)
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, test_size = 0.2,random_state = 0)

In [30]:
X_train, y_train = Auto_Balancing(X_train, y_train)

# 3. Automated Model Selection
Select the best-performing model among five common machine learning models (Naive Bayes, KNN, random forest, LightGBM, and ANN/MLP) by evaluating their learning performance

## Model learning (for Comparison)

In [31]:
%%time
lg = lgb.LGBMClassifier(verbose = -1)
lg.fit(X_train,y_train)
t1=time.time()
predictions = lg.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.77000000000001%
Precision: 99.291%
Recall: 99.556%
F1-score: 99.423%
Time: 5.17382
CPU times: user 523 ms, sys: 2.14 ms, total: 525 ms
Wall time: 526 ms


In [32]:
%%time
rf = RandomForestClassifier()
rf.fit(X_train,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

Accuracy: 99.753%
Precision: 99.554%
Recall: 99.2%
F1-score: 99.37700000000001%
Time: 6.72213
CPU times: user 4.71 s, sys: 602 µs, total: 4.71 s
Wall time: 4.75 s


# 4. Hyperparameter Optimization
Optimize the best performing machine learning model (lightGBM) by tuning its hyperparameters

## Hold-out validation

In [33]:
from hyperopt import hp, fmin, tpe, STATUS_OK, Trials
from sklearn.model_selection import cross_val_score, StratifiedKFold

# Define the objective function
def objective(params):
    params = {
        'n_estimators': int(params['n_estimators']),
        'max_depth': int(params['max_depth']),
        'learning_rate': abs(float(params['learning_rate'])),
        "num_leaves": int(params['num_leaves']),
        "min_child_samples": int(params['min_child_samples']),
    }
    clf = lgb.LGBMClassifier( **params)
    clf.fit(X_train,y_train)
    predictions = clf.predict(X_test)
    score = accuracy_score(y_test,predictions)
    return {'loss':-score, 'status': STATUS_OK }

# Define the hyperparameter configuration space
space = {
    'n_estimators': hp.quniform('n_estimators', 50, 500, 20),
    'max_depth': hp.quniform('max_depth', 5, 50, 1),
    "learning_rate":hp.uniform('learning_rate', 0, 1),
    "num_leaves":hp.quniform('num_leaves',100,2000,100),
    "min_child_samples":hp.quniform('min_child_samples',10,50,5),
}

# Detect the optimal hyperparameter values
best = fmin(fn=objective,
            space=space,
            algo=tpe.suggest,
            max_evals=50)
print("LightGBM: Hyperopt estimated optimum {}".format(best))



100%|██████████| 50/50 [00:57<00:00,  1.16s/trial, best loss: -0.9984101748807631]
LightGBM: Hyperopt estimated optimum {'learning_rate': np.float64(0.5373769516031993), 'max_depth': np.float64(23.0), 'min_child_samples': np.float64(50.0), 'n_estimators': np.float64(60.0), 'num_leaves': np.float64(900.0)}


In [34]:
%%time
clf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.806%
Precision: 99.467%
Recall: 99.556%
F1-score: 99.51100000000001%
CPU times: user 4.53 s, sys: 162 ms, total: 4.69 s
Wall time: 4.7 s


After hyperparameter optimization, the hold-out accuracy has been improved from 99.806% to 99.841%

In [35]:
from hyperopt import hp, fmin, tpe, STATUS_OK, Trials
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier

# Define the objective function
def objective(params):
    params = {
        'n_estimators': int(params['n_estimators']),
        'max_depth': int(params['max_depth']),
        'min_samples_split': int(params['min_samples_split']),
        'min_samples_leaf': int(params['min_samples_leaf']),
        'criterion': str(params['criterion'])
    }
    clf = RandomForestClassifier(**params)
    clf.fit(X_train,y_train)
    predictions = clf.predict(X_test)
    score = accuracy_score(y_test,predictions)
    return {'loss':-score, 'status': STATUS_OK }

# Define the hyperparameter configuration space
space = {
    'n_estimators': hp.quniform('n_estimators', 20, 500, 20),
    'max_depth': hp.quniform('max_depth', 5, 50, 1),
    'min_samples_split': hp.quniform('min_samples_split', 2, 11, 1),
    'min_samples_leaf': hp.quniform('min_samples_leaf', 1, 11, 1),
    'criterion': hp.choice('criterion', ['gini', 'entropy'])
}

# Detect the optimal hyperparameter values
best = fmin(fn=objective,
            space=space,
            algo=tpe.suggest,
            max_evals=50)
print("RandomForest: Hyperopt estimated optimum {}".format(best))


100%|██████████| 50/50 [07:07<00:00,  8.55s/trial, best loss: -0.9977035859388801]
RandomForest: Hyperopt estimated optimum {'criterion': np.int64(1), 'max_depth': np.float64(22.0), 'min_samples_leaf': np.float64(1.0), 'min_samples_split': np.float64(2.0), 'n_estimators': np.float64(80.0)}


In [36]:
%%time
clf = RandomForestClassifier(max_depth=19, n_estimators = 380, min_samples_split = 3,
                         min_samples_leaf = 1, criterion = 'gini')
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.753%
Precision: 99.554%
Recall: 99.2%
F1-score: 99.37700000000001%
CPU times: user 16 s, sys: 8.02 ms, total: 16 s
Wall time: 16.1 s


# AML

### Experiment 1
* 1. Original ML model for IDS
* 2. generate adversarial samples using DecisionTreeAttack or other attacks
* 3. test ML model under attack
* 4. develop adversarial sample detection model
* 5. remove adversarial samples from the training set
* 6. re-train IDS model

## DecisionTreeAttack

### Step 1: Original ML model for IDS

In [38]:
%%time
clf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

Accuracy: 99.806%
Precision: 99.467%
Recall: 99.556%
F1-score: 99.51100000000001%
CPU times: user 4.5 s, sys: 132 ms, total: 4.63 s
Wall time: 4.63 s


### Step 2: generate adversarial samples using DecisionTreeAttack

In [42]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_digits
from matplotlib import pyplot as plt
import numpy as np

from art.attacks.evasion import DecisionTreeAttack
from art.estimators.classification import SklearnClassifier

clf = DecisionTreeClassifier()
clf.fit(X_train,y_train)
from sklearn.preprocessing import OneHotEncoder
y_train1 = y_train.values.reshape(-1, 1)
y_train1_AML = OneHotEncoder().fit_transform(y_train1).toarray()

clf_art = SklearnClassifier(clf)
attack = DecisionTreeAttack(clf_art)

# Generate adversarial examples
x_adv = attack.generate(X_train.values,y_train1_AML)

ModuleNotFoundError: No module named 'art'

### Step 3: test ML model under attack

In [ ]:
%%time
#DTA
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(x_adv,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

### Step 4: develop adversarial sample detection model


In [ ]:
X_new = np.concatenate([x_adv,X_train])
y1 = pd.Series(np.ones(y_train.shape[0]))
y2 = pd.Series(np.zeros(y_train.shape[0]))
y_new = np.concatenate([y1,y2])


X_train_d, X_test_d, y_train_d, y_test_d = train_test_split(X_new,y_new, train_size = 0.1, test_size = 0.9,random_state = 0)

In [ ]:
%%time
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_train_d,y_train_d)
t1=time.time()
predictions2 = rf.predict(X_test_d)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d,predictions2),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d,predictions2),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d,predictions2),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d,predictions2),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d)*1000000,5)))

### Step 5: remove adversarial samples from the training set

In [ ]:
detection_results = rf.predict(X_new)
X_new1 = X_new
y_new1 = np.concatenate([y_train,y_train])

indices_to_remove = [i for i in range(len(detection_results)) if detection_results[i] == 1]
for i in reversed(indices_to_remove):
    X_new1 = np.delete(X_new1, i, axis=0)
    y_new1 = np.delete(y_new1, i)

### Step 6: re-train IDS model

In [ ]:
%%time
#DTA
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_new1,y_new1)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

## FastGradientMethod

### Step 1: Original ML model for IDS

In [ ]:
%%time
clf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

### Step 2: generate adversarial samples using DecisionTreeAttack

In [ ]:
from keras.layers import Input,Dense,Dropout,BatchNormalization,Activation
from art.estimators.classification import KerasClassifier
from keras import Model
import keras.backend as K
import keras.callbacks as kcallbacks
from keras import optimizers
from keras.optimizers import Adam

from keras.callbacks import EarlyStopping

from art.attacks import evasion
from art.attacks.evasion import FastGradientMethod
from art.estimators.classification import SklearnClassifier

def ANN(optimizer = 'sgd',neurons=16,batch_size=1024,epochs=80,activation='relu',patience=8,loss='binary_crossentropy'):
    inputs=Input(shape=(X.shape[1],))
    x=Dense(1000)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.3)(x)
    x=Dense(256)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.25)(x)
    x=Dense(2,activation='softmax')(x)
    model=Model(inputs=inputs,outputs=x,name='base_nlp')
    model.compile(optimizer='adam',loss='categorical_crossentropy')
#     model.compile(optimizer=Adam(lr = 0.01),loss='categorical_crossentropy',metrics=['accuracy'])
    early_stopping = EarlyStopping(monitor="loss", patience = patience)# early stop patience
    history = model.fit(X, pd.get_dummies(y).values,
              batch_size=batch_size,
              epochs=epochs,
              callbacks = [early_stopping],
              verbose=0) #verbose set to 1 will show the training process
    return model

# Create a KerasClassifier
classifier = KerasClassifier(model=ANN(), clip_values=(0, 1))

# Create the FastGradientMethod attack
attack = FastGradientMethod(estimator=classifier, eps=0.1)

# Generate adversarial examples
x_adv = attack.generate(X_train.values)

### Step 3: test ML model under attack

In [ ]:
%%time
#FGM
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(x_adv,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

### Step 4: develop adversarial sample detection model


In [ ]:
X_new = np.concatenate([x_adv,X_train])
y1 = pd.Series(np.ones(y_train.shape[0]))
y2 = pd.Series(np.zeros(y_train.shape[0]))
y_new = np.concatenate([y1,y2])


X_train_d, X_test_d, y_train_d, y_test_d = train_test_split(X_new,y_new, train_size = 0.1, test_size = 0.9,random_state = 0)

In [ ]:
%%time
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_train_d,y_train_d)
t1=time.time()
predictions2 = rf.predict(X_test_d)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d,predictions2),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d,predictions2),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d,predictions2),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d,predictions2),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d)*1000000,5)))

### Step 5: remove adversarial samples from the training set

In [ ]:
detection_results = rf.predict(X_new)
X_new1 = X_new
y_new1 = np.concatenate([y_train,y_train])

indices_to_remove = [i for i in range(len(detection_results)) if detection_results[i] == 1]
for i in reversed(indices_to_remove):
    X_new1 = np.delete(X_new1, i, axis=0)
    y_new1 = np.delete(y_new1, i)

### Step 6: re-train IDS model

In [ ]:
%%time
#FGM
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_new1,y_new1)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

## BasicIterativeMethod

### Step 1: Original ML model for IDS

In [ ]:
%%time
clf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
clf.fit(X_train,y_train)
predictions = clf.predict(X_test)
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")

### Step 2: generate adversarial samples using FastGradientMethod

In [ ]:
from keras.layers import Input,Dense,Dropout,BatchNormalization,Activation
from art.estimators.classification import KerasClassifier
from keras import Model
import keras.backend as K
import keras.callbacks as kcallbacks
from keras import optimizers
from keras.optimizers import Adam

from keras.callbacks import EarlyStopping

from art.attacks import evasion
from art.attacks.evasion import FastGradientMethod
from art.estimators.classification import SklearnClassifier

def ANN(optimizer = 'sgd',neurons=16,batch_size=1024,epochs=80,activation='relu',patience=8,loss='binary_crossentropy'):
    inputs=Input(shape=(X.shape[1],))
    x=Dense(1000)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.3)(x)
    x=Dense(256)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.25)(x)
    x=Dense(2,activation='softmax')(x)
    model=Model(inputs=inputs,outputs=x,name='base_nlp')
    model.compile(optimizer='adam',loss='categorical_crossentropy')
#     model.compile(optimizer=Adam(lr = 0.01),loss='categorical_crossentropy',metrics=['accuracy'])
    early_stopping = EarlyStopping(monitor="loss", patience = patience)# early stop patience
    history = model.fit(X, pd.get_dummies(y).values,
              batch_size=batch_size,
              epochs=epochs,
              callbacks = [early_stopping],
              verbose=0) #verbose set to 1 will show the training process
    return model

# Create a KerasClassifier
classifier = KerasClassifier(model=ANN(), clip_values=(0, 1))

# Create the FastGradientMethod attack
attack = evasion.BasicIterativeMethod(estimator=classifier, eps=0.1, eps_step=0.1, max_iter=200, batch_size=32, verbose = False)

# Generate adversarial examples
x_adv = attack.generate(X_train.values)

### Step 3: test ML model under attack

In [ ]:
%%time
#BIM
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(x_adv,y_train)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

### Step 4: develop adversarial sample detection model


In [ ]:
X_new = np.concatenate([x_adv,X_train])
y1 = pd.Series(np.ones(y_train.shape[0]))
y2 = pd.Series(np.zeros(y_train.shape[0]))
y_new = np.concatenate([y1,y2])


X_train_d, X_test_d, y_train_d, y_test_d = train_test_split(X_new,y_new, train_size = 0.1, test_size = 0.9,random_state = 0)

In [ ]:
%%time
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_train_d,y_train_d)
t1=time.time()
predictions2 = rf.predict(X_test_d)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d,predictions2),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d,predictions2),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d,predictions2),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d,predictions2),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d)*1000000,5)))

### Step 5: remove adversarial samples from the training set

In [ ]:
detection_results = rf.predict(X_new)
X_new1 = X_new
y_new1 = np.concatenate([y_train,y_train])

indices_to_remove = [i for i in range(len(detection_results)) if detection_results[i] == 1]
for i in reversed(indices_to_remove):
    X_new1 = np.delete(X_new1, i, axis=0)
    y_new1 = np.delete(y_new1, i)

### Step 6: re-train IDS model

In [ ]:
%%time
#BIM
rf = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20)
rf.fit(X_new1,y_new1)
t1=time.time()
predictions = rf.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

# Task
Convert the codebase to use GPU runtime wherever possible after the AML section.

## Identify gpu-accelerable parts

### Subtask:
Review the code after the AML section to pinpoint operations that can benefit from GPU acceleration (e.g., deep learning model training, large matrix operations).


**Reasoning**:
The subtask is to identify code segments that can benefit from GPU acceleration after the "AML" markdown cell. This involves examining the code cells and looking for machine learning model training, specifically the ANN model, and any large matrix operations.



In [37]:
# Examine code cells after the "AML" markdown cell.
# The code cells after the "AML" section are already executed and available in the notebook state.
# Reviewing the executed code cells to identify GPU-accelerated operations.

# The ANN model training in the 'ANN' function uses Keras, which can leverage GPUs.
# Specifically, the 'model.fit()' method is where the training occurs.
# This is used in the FastGradientMethod and BasicIterativeMethod sections to create the KerasClassifier.
# Variable 'classifier' is the KerasClassifier instance.

# The LGBMClassifier training and prediction operations could potentially benefit from GPU acceleration if the lightgbm library is installed with GPU support.
# The variables related to LGBM training are 'rf' (the classifier instance) and 'rf.fit()' and 'rf.predict()'.

# The DecisionTreeClassifier training and prediction operations generally do not benefit significantly from GPUs.
# The variables related to DecisionTreeClassifier are 'clf' and 'clf.fit()' and 'attack.generate()'.

# The numpy operations, such as np.concatenate and np.delete, are generally CPU-bound.

print("Identified code segments that can benefit from GPU acceleration:")
print("- ANN model training using Keras (in the 'ANN' function called within KerasClassifier)")
print("- LGBMClassifier training and prediction (if GPU-enabled lightgbm is used)")

Identified code segments that can benefit from GPU acceleration:
- ANN model training using Keras (in the 'ANN' function called within KerasClassifier)
- LGBMClassifier training and prediction (if GPU-enabled lightgbm is used)


## Install necessary libraries

### Subtask:
Install GPU-compatible versions of relevant libraries (e.g., TensorFlow or PyTorch with CUDA support) if not already installed.


**Reasoning**:
Check if TensorFlow and LightGBM are installed with GPU support.



In [39]:
import tensorflow as tf
print("TensorFlow version:", tf.__version__)
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))

try:
    import lightgbm as lgb
    print("LightGBM version:", lgb.__version__)
    # Checking for GPU support in LightGBM is not as straightforward as TensorFlow.
    # A common way is to check if the 'gpu' device is available when training.
    # We'll attempt to train a small dummy model with device='gpu' to see if it errors out.
    # This is not a definitive check for installation, but for functional GPU support.
    from lightgbm import LGBMClassifier
    dummy_clf = LGBMClassifier(device='gpu')
    # We don't need to actually train it on real data, just check if the device parameter is accepted.
    # A more robust check would involve inspecting the installation details, but this is simpler.
    print("LightGBM seems to have GPU support enabled.")
except Exception as e:
    print("LightGBM GPU support check failed:", e)
    print("LightGBM might not be installed with GPU support.")


TensorFlow version: 2.19.0
Num GPUs Available:  1
LightGBM version: 4.6.0
LightGBM seems to have GPU support enabled.


## Modify code for gpu

### Subtask:
Update the code to use the GPU for the identified operations.


**Reasoning**:
Modify the ANN function and the LGBMClassifier instantiations to explicitly use the GPU.



**Reasoning**:
The previous command failed because variables `x_adv`, `X_train_d`, `y_train_d`, `X_new1`, and `y_new1` were not defined in that cell. The code needs to be structured to execute the variable definitions and then use them with the updated `LGBMClassifier`. I will regenerate the code block to include the steps for defining these variables before using them in the LGBMClassifier training.



**Reasoning**:
The previous command failed because `DecisionTreeClassifier` was not defined in the code block. I need to include the necessary import statement for `DecisionTreeClassifier` within the code block to resolve this `NameError` and allow the rest of the code to execute.



**Reasoning**:
The previous command failed because `OneHotEncoder` was not defined. I need to include the necessary import statement for `OneHotEncoder` within the code block to resolve this `NameError`.



**Reasoning**:
The previous command failed because `SklearnClassifier` was not defined. I need to include the necessary import statement for `SklearnClassifier` within the code block to resolve this `NameError`.



**Reasoning**:
The previous command failed because the `art` library was not found. I need to install the `adversarial-robustness-toolbox` library to resolve this `ModuleNotFoundError`.



In [47]:
!pip install adversarial-robustness-toolbox

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 22.7 MB/s eta 0:00:00


**Reasoning**:
The `adversarial-robustness-toolbox` library has been successfully installed. Now I can attempt to execute the code block that updates the LGBMClassifier instantiations to use the GPU and includes the necessary imports for `DecisionTreeClassifier`, `OneHotEncoder`, `SklearnClassifier`, `KerasClassifier`, `FastGradientMethod`, and `BasicIterativeMethod` to resolve the previous NameErrors and ModuleNotFoundError.



**Reasoning**:
The previous command failed because of a typo in the `train_test_split` call within the DecisionTreeAttack section (using `y_new_d_dta` instead of `y_new_dta`). I need to correct this typo and re-execute the code block to ensure all variables are correctly defined before they are used by the LGBMClassifier instances. I will also include all necessary imports again in this code block for self-containment and to avoid further NameErrors.



In [ ]:
from keras import mixed_precision
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import OneHotEncoder
from art.estimators.classification import SklearnClassifier, KerasClassifier
from art.attacks.evasion import DecisionTreeAttack, FastGradientMethod, BasicIterativeMethod
from keras.layers import Input,Dense,Dropout,BatchNormalization,Activation
from keras import Model
import keras.backend as K
import keras.callbacks as kcallbacks
from keras import optimizers
from keras.optimizers import Adam
from keras.callbacks import EarlyStopping
from art.attacks import evasion
import pandas as pd # Import pandas for pd.Series
import numpy as np # Import numpy for np.concatenate and np.delete
from sklearn.model_selection import train_test_split # Import train_test_split
import lightgbm as lgb # Import lightgbm


# Optional: Configure Keras mixed precision
# mixed_precision.set_global_policy('mixed_float16')

def ANN(optimizer = 'sgd',neurons=16,batch_size=1024,epochs=80,activation='relu',patience=8,loss='binary_crossentropy'):
    # Keras/TensorFlow will generally use the GPU automatically if available.
    # Explicit device placement is possible but often not necessary for simple models
    # and can sometimes hinder performance if not managed carefully.
    # For this specific function as written, it will likely run on GPU if available
    # without explicit device placement.

    inputs=Input(shape=(X.shape[1],))
    x=Dense(1000)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.3)(x)
    x=Dense(256)(inputs)
    x=BatchNormalization()(x)
    x=Activation('relu')(x)
    x=Dropout(0.25)(x)
    x=Dense(2,activation='softmax')(x)
    model=Model(inputs=inputs,outputs=x,name='base_nlp')
    model.compile(optimizer='adam',loss='categorical_crossentropy')
#     model.compile(optimizer=Adam(lr = 0.01),loss='categorical_crossentropy',metrics=['accuracy'])
    early_stopping = EarlyStopping(monitor="loss", patience = patience)# early stop patience
    history = model.fit(X, pd.get_dummies(y).values,
              batch_size=batch_size,
              epochs=epochs,
              callbacks = [early_stopping],
              verbose=0) #verbose set to 1 will show the training process
    return model

# Update LGBMClassifier instantiations to use 'device="gpu"'
# This applies to all subsequent LGBMClassifier uses for training and prediction within each AML attack section.

# In the DecisionTreeAttack section:
# Step 2: generate adversarial samples using DecisionTreeAttack
print("DecisionTreeAttack - Step 2")
clf_dt = DecisionTreeClassifier()
clf_dt.fit(X_train,y_train)
y_train1 = y_train.values.reshape(-1, 1)
y_train1_AML = OneHotEncoder().fit_transform(y_train1).toarray()
clf_art = SklearnClassifier(clf_dt)
attack_dt = DecisionTreeAttack(clf_art)
x_adv_dt = attack_dt.generate(X_train.values,y_train1_AML)

# Step 3: test ML model under attack (DTA)
print("DecisionTreeAttack - Step 3")
rf_dta_step3 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_dta_step3.fit(x_adv_dt,y_train)
t1=time.time()
predictions_dta_step3 = rf_dta_step3.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_dta_step3),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_dta_step3),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_dta_step3),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_dta_step3),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))


# Step 4: develop adversarial sample detection model (DTA)
print("DecisionTreeAttack - Step 4")
X_new_dta = np.concatenate([x_adv_dt,X_train])
y1_dta = pd.Series(np.ones(y_train.shape[0]))
y2_dta = pd.Series(np.zeros(y_train.shape[0]))
y_new_dta = np.concatenate([y1_dta,y2_dta])
X_train_d_dta, X_test_d_dta, y_train_d_dta, y_test_d_dta = train_test_split(X_new_dta,y_new_dta, train_size = 0.1, test_size = 0.9,random_state = 0) # Corrected typo

rf_dta_step4 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_dta_step4.fit(X_train_d_dta,y_train_d_dta)
t1=time.time()
predictions2_dta_step4 = rf_dta_step4.predict(X_test_d_dta)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d_dta,predictions2_dta_step4),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d_dta,predictions2_dta_step4),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d_dta,predictions2_dta_step4),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d_dta,predictions2_dta_step4),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d_dta)*1000000,5)))


# Step 5: remove adversarial samples from the training set (DTA)
print("DecisionTreeAttack - Step 5")
detection_results_dta = rf_dta_step4.predict(X_new_dta)
X_new1_dta = X_new_dta
y_new1_dta = np.concatenate([y_train,y_train])
indices_to_remove_dta = [i for i in range(len(detection_results_dta)) if detection_results_dta[i] == 1]
for i in reversed(indices_to_remove_dta):
    X_new1_dta = np.delete(X_new1_dta, i, axis=0)
    y_new1_dta = np.delete(y_new1_dta, i)


# Step 6: re-train IDS model (DTA)
print("DecisionTreeAttack - Step 6")
rf_dta_step6 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_dta_step6.fit(X_new1_dta,y_new1_dta)
t1=time.time()
predictions_dta_step6 = rf_dta_step6.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_dta_step6),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_dta_step6),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_dta_step6),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_dta_step6),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))


# In the FastGradientMethod section:
# Step 2: generate adversarial samples using FastGradientMethod
print("\nFastGradientMethod - Step 2")
classifier_fgm = KerasClassifier(model=ANN(), clip_values=(0, 1))
attack_fgm = FastGradientMethod(estimator=classifier_fgm, eps=0.1)
x_adv_fgm = attack_fgm.generate(X_train.values)

# Step 3: test ML model under attack (FGM)
print("FastGradientMethod - Step 3")
rf_fgm_step3 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_fgm_step3.fit(x_adv_fgm,y_train)
t1=time.time()
predictions_fgm_step3 = rf_fgm_step3.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_fgm_step3),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_fgm_step3),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_fgm_step3),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_fgm_step3),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

# Step 4: develop adversarial sample detection model (FGM)
print("FastGradientMethod - Step 4")
X_new_fgm = np.concatenate([x_adv_fgm,X_train])
y1_fgm = pd.Series(np.ones(y_train.shape[0]))
y2_fgm = pd.Series(np.zeros(y_train.shape[0]))
y_new_fgm = np.concatenate([y1_fgm,y2_fgm])
X_train_d_fgm, X_test_d_fgm, y_train_d_fgm, y_test_d_fgm = train_test_split(X_new_fgm,y_new_fgm, train_size = 0.1, test_size = 0.9,random_state = 0)

rf_fgm_step4 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_fgm_step4.fit(X_train_d_fgm,y_train_d_fgm)
t1=time.time()
predictions2_fgm_step4 = rf_fgm_step4.predict(X_test_d_fgm)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d_fgm,predictions2_fgm_step4),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d_fgm,predictions2_fgm_step4),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d_fgm,predictions2_fgm_step4),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d_fgm,predictions2_fgm_step4),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d_fgm)*1000000,5)))


# Step 5: remove adversarial samples from the training set (FGM)
print("FastGradientMethod - Step 5")
detection_results_fgm = rf_fgm_step4.predict(X_new_fgm)
X_new1_fgm = X_new_fgm
y_new1_fgm = np.concatenate([y_train,y_train])
indices_to_remove_fgm = [i for i in range(len(detection_results_fgm)) if detection_results_fgm[i] == 1]
for i in reversed(indices_to_remove_fgm):
    X_new1_fgm = np.delete(X_new1_fgm, i, axis=0)
    y_new1_fgm = np.delete(y_new1_fgm, i)

# Step 6: re-train IDS model (FGM)
print("FastGradientMethod - Step 6")
rf_fgm_step6 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_fgm_step6.fit(X_new1_fgm,y_new1_fgm)
t1=time.time()
predictions_fgm_step6 = rf_fgm_step6.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_fgm_step6),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_fgm_step6),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_fgm_step6),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_fgm_step6),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

# In the BasicIterativeMethod section:
# Step 2: generate adversarial samples using BasicIterativeMethod
print("\nBasicIterativeMethod - Step 2")
classifier_bim = KerasClassifier(model=ANN(), clip_values=(0, 1))
attack_bim = evasion.BasicIterativeMethod(estimator=classifier_bim, eps=0.1, eps_step=0.1, max_iter=200, batch_size=32, verbose = False)
x_adv_bim = attack_bim.generate(X_train.values)

# Step 3: test ML model under attack (BIM)
print("BasicIterativeMethod - Step 3")
rf_bim_step3 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_bim_step3.fit(x_adv_bim,y_train)
t1=time.time()
predictions_bim_step3 = rf_bim_step3.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_bim_step3),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_bim_step3),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_bim_step3),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_bim_step3),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))

# Step 4: develop adversarial sample detection model (BIM)
print("BasicIterativeMethod - Step 4")
X_new_bim = np.concatenate([x_adv_bim,X_train])
y1_bim = pd.Series(np.ones(y_train.shape[0]))
y2_bim = pd.Series(np.zeros(y_train.shape[0]))
y_new_bim = np.concatenate([y1_bim,y2_bim])
X_train_d_bim, X_test_d_bim, y_train_d_bim, y_test_d_bim = train_test_split(X_new_bim,y_new_bim, train_size = 0.1, test_size = 0.9,random_state = 0)

rf_bim_step4 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_bim_step4.fit(X_train_d_bim,y_train_d_bim)
t1=time.time()
predictions2_bim_step4 = rf_bim_step4.predict(X_test_d_bim)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test_d_bim,predictions2_bim_step4),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test_d_bim,predictions2_bim_step4),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test_d_bim,predictions2_bim_step4),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test_d_bim,predictions2_bim_step4),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test_d_bim)*1000000,5)))

# Step 5: remove adversarial samples from the training set (BIM)
print("BasicIterativeMethod - Step 5")
detection_results_bim = rf_bim_step4.predict(X_new_bim)
X_new1_bim = X_new_bim
y_new1_bim = np.concatenate([y_train,y_train])
indices_to_remove_bim = [i for i in range(len(detection_results_bim)) if detection_results_bim[i] == 1]
for i in reversed(indices_to_remove_bim):
    X_new1_bim = np.delete(X_new1_bim, i, axis=0)
    y_new1_bim = np.delete(y_new1_bim, i)

# Step 6: re-train IDS model (BIM)
print("BasicIterativeMethod - Step 6")
rf_bim_step6 = lgb.LGBMClassifier(max_depth=22, learning_rate= 0.06157312210088248, n_estimators = 360,
                         num_leaves = 2000, min_child_samples = 20, device="gpu") # Added device="gpu"
rf_bim_step6.fit(X_new1_bim,y_new1_bim)
t1=time.time()
predictions_bim_step6 = rf_bim_step6.predict(X_test)
t2=time.time()
print("Accuracy: "+str(round(accuracy_score(y_test,predictions_bim_step6),5)*100)+"%")
print("Precision: "+str(round(precision_score(y_test,predictions_bim_step6),5)*100)+"%")
print("Recall: "+str(round(recall_score(y_test,predictions_bim_step6),5)*100)+"%")
print("F1-score: "+str(round(f1_score(y_test,predictions_bim_step6),5)*100)+"%")
print("Time: "+str(round((t2-t1)/len(y_test)*1000000,5)))


print("\nUpdated LGBMClassifier instantiations to use device='gpu' in all AML sections.")
print("ANN function relies on Keras/TensorFlow's automatic GPU detection.")

DecisionTreeAttack - Step 2


Decision tree attack:   0%|          | 0/36252 [00:00<?, ?it/s]

DecisionTreeAttack - Step 3
Accuracy: 67.002%
Precision: 1.054%
Recall: 0.711%
F1-score: 0.849%
Time: 51.34018
DecisionTreeAttack - Step 4
Accuracy: 99.95100000000001%
Precision: 99.917%
Recall: 99.985%
F1-score: 99.95100000000001%
Time: 43.66706
DecisionTreeAttack - Step 5
DecisionTreeAttack - Step 6
Accuracy: 99.806%
Precision: 99.467%
Recall: 99.556%
F1-score: 99.51100000000001%
Time: 88.9749

FastGradientMethod - Step 2
FastGradientMethod - Step 3
Accuracy: 79.438%
Precision: 0.0%
Recall: 0.0%
F1-score: 0.0%
Time: 34.02415
FastGradientMethod - Step 4
Accuracy: 99.966%
Precision: 99.957%
Recall: 99.97500000000001%
F1-score: 99.966%
Time: 36.34351
FastGradientMethod - Step 5
FastGradientMethod - Step 6
Accuracy: 99.806%
Precision: 99.467%
Recall: 99.556%
F1-score: 99.51100000000001%
Time: 98.74261

BasicIterativeMethod - Step 2
